### PR TITLE
Refactoring Windows certificate code and tests to reflect updates in the win32-certstore gem

### PR DESCRIPTION
**Pull request by johnmccrae**
Signed-off-by: John McCrae <jmccrae@chf.io>

Updated windows_certificate resource to account for bugs and updates in the win32-certstore gem.

## Description
We discovered problems with the win32-certstore gem not finding and returning certificates correctly. We corrected that and as a consequence, the spec/functional/resource/windows_certificate_spec.rb tests were failing. Those are updated to account for that. We took a softer approach and opted to log certificate failures rather than throwing assertions which was requested by the community. 

Additionally, the kitchen.yml, macos.rb and a certificate file in the /kitchen-test folder were updated to account for problems in the build test process for the end-to-end cookbook.

Finally, there was a small, completely unrelated change made to the /lib/chef/http/authenticator.rb file. It has code in it to randomly generate a password used to secure certificates with. That password generator code was not consistently creating a password of the desired length of 14 characters. The result was that ~15% of the time, unit tests on Windows would fail and kill the build. I updated that code to be more explicit about the password length. 

Additionally, 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: omnibus/Gemfile.lock [hunk 1]
```diff
@@ -1,6 +1,6 @@
 GIT
   remote: https://github.com/chef/omnibus-software.git
-  revision: 8a0e391d4232ab3fc5fb7a03b15932335a0214c4
```
Comment (neha-p6) [inline omnibus/Gemfile.lock:3]: @johnmccrae Should we also create and see adhoc build passes before merging, due to updates in omnibus/Gemfile.lock?

**Comment by neha-p6** [comment]
@johnmccrae  Have you seen kitchen fedora failure? 
Also we would have to backport it to chef-17? 
It all started with this issue https://github.com/chef/chef/issues/11994 needing a fix as far as I remember

File: kitchen-tests/cookbooks/end_to_end/recipes/macos.rb [hunk 2]
```diff
@@ -73,25 +73,15 @@
...
-
-execute "changing ownership of the git cask" do
-  command "chmod -R 777 /usr/local/Cellar/git@2.35.1"
-  live_stream true
-end
-
-execute "Sledge Hammer removal of the offending Git version" do
-  command "rm -rf /usr/local/Cellar/git@2.35.1"
-  live_stream true
-end
```
Comment (marcparadise) [inline kitchen-tests/cookbooks/end_to_end/recipes/macos.rb:94]: Is this related tot his change?  If not, it should be in its own PR.

File: lib/chef/resource/windows_certificate.rb [hunk 3]
```diff
@@ -128,14 +128,14 @@ class WindowsCertificate < Chef::Resource
       end
 
       action :delete, description: "Deletes a certificate." do
-        cert_obj = fetch_cert
+        cert_obj = verify_cert
 
-        if cert_obj
+        if cert_obj == true || cert_obj == "Certificate Has Expired"
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:133]: `verify_cert` only returns a `true` or `false`, so this condition can never be met. 

Given that, can we change `cert_obj` name to something like `cert_is_valid`?

File: lib/chef/resource/windows_certificate.rb [hunk 4]
```diff
@@ -150,12 +150,12 @@ class WindowsCertificate < Chef::Resource
           cert_obj = fetch_cert
         end
 
-        if cert_obj
+        if cert_obj != "Certificate Not Found"
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:153]: As above, this will always be true because cert_obj can only be true/false.

File: lib/chef/resource/windows_certificate.rb [hunk 5]
```diff
@@ -261,9 +256,11 @@ def get_thumbprint(store_name, location, source)
         end
 
         def resolve_thumbprint(thumbprint)
-          return thumbprint if valid_thumbprint?(thumbprint)
-
-          powershell_exec!(get_thumbprint(new_resource.store_name, ps_cert_location, new_resource.source)).result
+          # valid_thumbprint can return false under at least 2 conditions:
+          # one is that the thumbprint is in fact busted
+          # the second is that the thumbprint is valid but belongs to an expired certificate already installed
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:261]: This comment does not seem accurate.  Above, `valid_thumbprint?` is defined as only: 
```
 string.match?(/[0-9A-Fa-f]/) && string.length == 40
```
`resolve_thumbprint` in its current form only validates the correctness of the thumbprint.  Is that the  intent?

File: lib/chef/resource/windows_certificate.rb [hunk 6]
```diff
@@ -275,14 +272,11 @@ def resolve_thumbprint(thumbprint)
 
         def verify_cert(thumbprint = new_resource.source)
           store = ::Win32::Certstore.open(new_resource.store_name, store_location: native_cert_location)
-          if new_resource.pfx_password.nil?
-            store.valid?(resolve_thumbprint(thumbprint), store_location: native_cert_location, store_name: new_resource.store_name )
-          else
-            store.valid?(resolve_thumbprint(thumbprint), store_location: native_cert_location, store_name: new_resource.store_name)
-          end
+          result = store.valid?(resolve_thumbprint(thumbprint))
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:275]: What does `store.valid?` do when `resolve_thumbprint` returns `false`?

File: lib/chef/resource/windows_certificate.rb [hunk 7]
```diff
@@ -496,20 +488,15 @@ def import_certificates(cert_objs, is_pfx, store_name: new_resource.store_name,
                   converge_by("Creating a PFX #{@local_pfx_path} for Store #{new_resource.store_name}") do
                     add_pfx_cert(@local_pfx_path)
                   end
-                else
-                  message = "You passed an invalid file or url to import. Please check the spelling and try again."
-                  message << exception.message
-                  raise Chef::Exceptions::ArgumentError, message
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:502]: Why shouldn't we raise here? The resource can't converge in this condition, right?

File: lib/chef/resource/windows_certificate.rb (hunk 6 above)
Comment (marcparadise) [reply lib/chef/resource/windows_certificate.rb:275]: Separately, why is a method named `valid?` maybe-returning a string?

File: lib/chef/resource/windows_certificate.rb [hunk 8]
```diff
@@ -480,14 +474,12 @@ def export_cert(cert_obj, output_path:, store_name:, store_location:, pfx_passwo
...
           [cert_objs].flatten.each do |cert_obj|
-            # thumbprint = OpenSSL::Digest.new("SHA1", cert_obj.to_der).to_s
-            # pkcs = OpenSSL::PKCS12.new(cert_obj, new_resource.pfx_password)
-            # cert = OpenSSL::X509::Certificate.new(pkcs.certificate.to_pem)
             thumbprint = OpenSSL::Digest.new("SHA1", cert_obj.to_der).to_s
-            if is_pfx
-              if verify_cert(thumbprint) == true
-                Chef::Log.debug("Certificate is already present")
-              else
+            results = verify_cert(thumbprint)
```
Comment (marcparadise) [inline lib/chef/resource/windows_certificate.rb:478]: Seems like we could stick to `if verify_cert(thumbprint)....else...` instead of assigning to `results` and also checking `== false` (since verify_cert returns only true/false)

**Review by marcparadise — CHANGES_REQUESTED** [review]

File: omnibus/Gemfile.lock (hunk 1 above)
Comment (johnmccrae) [reply omnibus/Gemfile.lock:3]: Great point. I'll roll one after I update with Marc's comments

File: kitchen-tests/cookbooks/end_to_end/recipes/macos.rb (hunk 2 above)
Comment (johnmccrae) [reply kitchen-tests/cookbooks/end_to_end/recipes/macos.rb:94]: ok, thanks. Pulling that out.

File: lib/chef/resource/windows_certificate.rb (hunk 3 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:133]: ooh, good catch. Thanks. Done and done

File: lib/chef/resource/windows_certificate.rb (hunk 4 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:153]: In this case, the fetch method DOES return a Not Found

Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:153]: The fetch method ultimately calls get_cert from win32-cerstore which can return a Cert Not Found

Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:153]: I did update this though since it was wrong

File: lib/chef/resource/windows_certificate.rb (hunk 5 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:261]: I updated that method name to validate_thumbprint.

File: lib/chef/resource/windows_certificate.rb (hunk 6 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:275]: We check that the format of the thumbprint is correct - 40 hex characters with no spaces. If that fails, we throw an exception now (just added that in). Otherwise, we search for a cert that matches that borked thumbprint and return "not found"

File: lib/chef/resource/windows_certificate.rb (hunk 7 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:502]: oopsy, that was a mistake. Added that back in

File: lib/chef/resource/windows_certificate.rb (hunk 8 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:478]: I can see that. I was trying to simplify this. Rolled back a bit here.
